react-renderer/src/Optimizer.ts

```diff
@@ -213,72 +215,75 @@ export const stepEP = (state: State, steps: number, evaluate = true) => {
+        const constrEng: Tensor =
+            constrEngs.length === 0 ? differentiable(0) : stack(constrEngs).sum();
+        const overallEng = objEng.add(
+            constrEng.mul(scalar(constraintWeight * state.params.weight))
+        );
+
+        // NOTE: the current version of tfjs requires all input variables to have gradients (i.e. actually involved when computing the overall energy). See https://github.com/tensorflow/tfjs-core/blob/8c2d9e05643988fa7f4575c30a5ad3e732d189b2/tfjs-core/src/engine.ts#L726
+        // HACK: therefore, we try to work around it by using all varying values without affecting the value and gradients of the energy function
+        const dummyVal = stack(varyingValuesTF).sum();
+        return overallEng.add(dummyVal.mul(scalar(0)));
+    };
 };
 
+// TODO: This is an old function and should be phased out; may no longer compile
 export const step = (state: State, steps: number) => {
-  const f = evalEnergyOn(state);
-  const fgrad = gradF(f);
-  const xs = state.varyingValues.map(differentiable);
-  // const xs = state.varyingState; // NOTE: use cached varying values
-  // NOTE: minimize will mutates xs
-  const { energy } = minimize(f, fgrad, xs, steps);
-  // insert the resulting variables back into the translation for rendering
-  // NOTE: this is a synchronous operation on all varying values; may block
-  const varyingValues = xs.map((x) => scalarValue(x as Scalar));
-  const trans = insertVaryings(
-    state.translation,
-    zip(state.varyingPaths, varyingValues) as [Path, number][]
-  );
-  const newState = { ...state, translation: trans, varyingValues };
-  if (scalarValue(energy) > 10) {
-    // const newState = { ...state, varyingState: xs };
-    newState.params.optStatus.tag = "UnconstrainedRunning";
-    console.log(`Took ${steps} steps. Current energy`, scalarValue(energy));
-    // return newState;
-  } else {
-    // const varyingValues = xs.map((x) => tfStr(x));
-    // const trans = insertVaryings(
-    //   state.translation,
-    //   zip(state.varyingPaths, varyingValues) as [Path, number][]
-    // );
-    // const newState = { ...state, translation: trans, varyingValues };
-    newState.params.optStatus.tag = "EPConverged";
-    console.log("Converged with energy", scalarValue(energy));
-    // return evalTranslation(newState);
-  }
-  // return the state with a new set of shapes
-  return evalTranslation(newState);
+    const f = evalEnergyOn(state);
+    const fgrad = gradF(f);
+    const xs = state.varyingValues;
+    // const xs = state.varyingState; // NOTE: use cached varying values
+    // NOTE: minimize will mutates xs
+    const { energy } = minimize(f, fgrad, xs, steps);
+    // insert the resulting variables back into the translation for rendering
+    // NOTE: this is a synchronous operation on all varying values; may block
+    // Note: after we finish one evaluation, we "destroy" the AD variables (by making them into scalars) and remake them
```

Comment:
What's the consequence of not "destroying" the AD variables?

Reply:
If you reuse an AD variable later with our dynamic evaluator approach, then my guess is that Tensorflow records those additional operations in its computational graph, which might affect the gradient. But it's a moot point after we convert to using our custom AD + compiled gradients, because the AD vars become irrelevant after the gradient is compiled.